### ISSUE TITLE
Работа со структурами С++

### ISSUE THREAD

**Issue by Geniok** [Pahtus200Maks0/Test#6]
Если рассмотреть такой код, то почему для структуры были выбраны именно массивы ?

> const int SLEN = 30;

struct student
{
    char fullname[SLEN];
    char hobby[SLEN];
    int ooplevel1;
};

Если рассмотреть данную функцию, за что она отвечает ?

> int getinfo(student ptr[], int clas)
{
    int coun = 0;
    std::cout << "\nВведите информацию о " << clas << " студентах\n\n";
    for (int i = 0; i < clas; ++i)
    {
        std::cout << "Введите имя студента N" << i + 1 << ' ';
        std::cin.getline(ptr[i].fullname, SLEN);
        if (strlen(ptr[i].fullname) == 0)
            break;
        std::cout << "\nВведите его хобби ";
        std::cin.getline(ptr[i].hobby, SLEN);
        std::cout << "\nВведите его уровень ";
        std::cin >> ptr[i].ooplevel1;
        std::cin.get();
        std::cout << '\n';
        coun++;
    }
    return coun;
}

Почему фамилия проверяется на нулевую длину, а хобби и уровень нет ?
Получается, фамилию я пропустить не могу, но все остальное могу ?
Тогда какие данные попадут в структуру ?

Для чего в данной функции используется переменная coun ?

**Comment by Pahtus200Maks0** [comment]
1 - Если рассмотреть такой код, то почему для структуры были выбраны именно массивы - просто как максимально знакомый мне вариант, но в лучше было бы использовать  объект string

2- Если рассмотреть данную функцию, за что она отвечает ? - выделяет память, инициализирует структуры, запрашивает ввод.

3. Почему фамилия проверяется на нулевую длину, а хобби и уровень нет ? - ну это просто недоработка, надо либо уведомить пользователя либо проверять на нулевую длину все. 

4 Получается, фамилию я пропустить не могу, но все остальное могу? - да, так. Недоработан этот момент.

5. Тогда какие данные попадут в структуру? - если не инициализировать член структуры, то он автоматом инициализируется в ноль.(но ели я ничего не путаю, то это зависит от компилятора, может и какой нибудь мусор туда положить).

**Comment by Pahtus200Maks0** [comment]
Для чего в данной функции используется переменная coun? - что бы инициализировать переменную  int entered и отобразить столько структур, сколько инициализировал пользователь. Это просто счётчик

**Comment by Geniok** [comment]
> Для чего в данной функции используется переменная coun? - что бы инициализировать переменную int entered и отобразить столько структур, сколько инициализировал пользователь. Это просто счётчик

А зачем, если вы уже это число знаете. запросив его ранее ?
Вы предполагаете, что возможна ситуация, когда пользователь на ваш запрос, "Какое количество структур Вы хотите инициализировать"  введет 30, а coun будет равно по выходе из функции getinfo() 15 ?
На мой взгляд количество структур вам известно еще ДО вызова функции getinfo. Какой смысл их подсчитывать еще раз в функции getinfo() ?
Возможно я ошибаюсь, но пока как мне кажется вы дважды получаете информацию, которой уже владеете, ввода как минимум две лишние переменные.

**Comment by Geniok** [comment]
> 1 - Если рассмотреть такой код, то почему для структуры были выбраны именно массивы - просто как максимально знакомый мне вариант, но в лучше было бы использовать объект string

Вы же написали что отлично знаете С++, но при этом не используете один из основных типов данных С++.


> 
> 2- Если рассмотреть данную функцию, за что она отвечает ? - выделяет память, инициализирует структуры, запрашивает ввод.

Ну то есть у вас функция отвечает за все сразу и заменяет по сути целую программу. Я бы посоветовал почитать о принципах SOLID.
> 
> 5. Тогда какие данные попадут в структуру? - если не инициализировать член структуры, то он автоматом инициализируется в ноль.(но ели я ничего не путаю, то это зависит от компилятора, может и какой-нибудь мусор туда положить).

Что не очень хорошо, согласны, когда даже вы сами как автор не уверены в поведении своей программы.
Хотя если ее чуть-чуть изменить, то поведение будет явное и вам не придется гадать, ка кона будет себя вести с разными компиляторами.

**Comment by Pahtus200Maks0** [comment]
1 - А зачем, если вы уже это число знаете. запросив его ранее ? - Тут суть в том, что если пользователь запросит 30 структур, а инициализирует только 1, то благодаря счётчику coun, код выведет на экран только 1 инициализированную структуру.
Счётчик coun нужен обязательно - он содержит количество фактически инициализированных структур. 

2 - Основная проблема вашего подхода заключается в том, что вам приходится память выделать - Да, тут код надо вообще переписывать и выделять память только под инициализированные структуры.

3 -  Что не очень хорошо, согласны? - Конечно не очень хорошо.

**Comment by Geniok** [comment]
>Тут суть в том, что если пользователь запросит 30 структур, а инициализирует только 1, то благодаря счётчику coun, код выведет на экран только 1 инициализированную структуру. Счётчик coun нужен обязательно - он содержит количество фактически инициализированных структур.

Так, погодите. Значит у вас программа работает следующим образом, пользователь хочет инициализировать 5 элементов.
Вводит 1 элемент.
На 2-м что он должен сделать, чтобы 2-й элемент не был инициализирован ?

**Comment by Pahtus200Maks0** [comment]
Пользователь захотел инициализировать 5 структур, а потом передумал и инициализировал только 1.
На 2-м, он в поле имени просто нажимает enter, генерируя \n. И программа отобразит именно 1 инициализированную структуру, потому что coun = 1.

Программа конечно реализована для пользователя очень не понятна. Но суть в том, что бы отобразить именно то количество структур, которое было инициализировано, а не то, которое забил пользователь изначально.

**Comment by Geniok** [comment]
> Пользователь захотел инициализировать 5 структур, а потом передумал и инициализировал только 1. На 2-м, он в поле имени просто нажимает enter, генерируя \n. И программа отобразит именно 1 инициализированную структуру, потому что coun = 1.
> 
> Программа конечно реализована для пользователя очень не понятна. Но суть в том, что бы отобразить именно то количество структур, которое было инициализировано, а не то, которое забил пользователь изначально.

Извините конечно, но это вы сами придумали, что он нажмет Энтер.
Пользователь введя изначально число 5, всегда будет заполнять все 5 структур. Но нажав **случайно** энтер, будет очень сильно удивлен результату. Но результат этот не запланирован изначально, он связан с тем, как написана программа.

Давайте представим, что таким образом написана программа банкомата, где вы снимаете деньги.
Вы подходите к банкомату, видите диалог, где вам предлагают ввести желаемую сумму. Вы вводите число 1000. Банкомат начинает отсчитывать нужную вам сумму. Вы ожидаете, что вам банкомат выдаст 1000 рублей, так ?

Но по вышей логике, программист, который писал ПО для банкомата, ввел в ПО следующую фишку: если вы **случайно** в процессе отсчета банкоматом денет нажмете на клавиатуре цифру 5, то он вам выдаст не всю сумму, а только ту, что уже успел отсчитать.
И вот вы пишите в техподдержку банка, узнать, а что случилось, и вам объясняют, что вот программист придумал такую фишку.
Думаю скорее всего свой банк после этого вы смените. 

Если бы в диалоге с пользователем были прописаны ВСЕ возможные действия, был бы обработчик неверных действий, были бы учтены все возможные варианты, тогда я бы с вами согласился. Но пока это выглядит как баг, и код построен вокруг этого бага.

**Comment by Pahtus200Maks0** [comment]
Вы проверяете свое ПО при тестировании как программист, зная исходники, а вам нужно проверять как пользователь, не зная, как оно реализовано внутри. - Да, понятно, возьму на заметку.
